---
title: "T analysis limb_mesoderm"
output:
  html_document:
    number_sections: yes
    toc: yes  
    keep_md: yes
---

# Set-up and loading data and intermediate results

```{r, echo=FALSE, output=FALSE, message=FALSE}
source("core_functions_dynamic.R")
source("core_functions_DA_testing_general.R")
source("chimera_core_functions_big_atlas.R")
library(ggplot2)
library(Polychrome)
library(progeny)
library(ggrepel)
library(progeny)
library(MAYA)

chimeraWT <- readRDS("../results/chimeraWT_2020_extended_mapping_rev.rds")
chimeraT <- readRDS("../results/chimeraT_2020_extended_mapping_rev.rds")
correlation_pseudotime_WT <- readRDS("results_no_split/correlation_pseudotime_WT.rds")
correlation_pseudotime_T <- readRDS("results_no_split/correlation_pseudotime_T.rds")
Celltypes <- read.table("lineage_names.txt")$V1
names(correlation_pseudotime_WT) <- Celltypes
names(correlation_pseudotime_T) <- Celltypes
correlation_pseudotime_WT <- correlation_pseudotime_WT[unlist(lapply(correlation_pseudotime_WT, function(x) !(is.null(x))))]
correlation_pseudotime_T <- correlation_pseudotime_T[unlist(lapply(correlation_pseudotime_T, function(x) !(is.null(x))))]

set.seed(44)
knitr::opts_chunk$set(
    echo = FALSE,
    message = FALSE,
    warning = FALSE,
    message = FALSE,
    dev = c("pdf"),
    dpi=300
)
options(datatable.fread.datatable=FALSE)
library(SingleCellExperiment)
library(data.table)
library(destiny)
library(scran)
library(scater)
library(batchelor)
```

# Remapping celltype and stage using correlation-based analysis on the specific trajectory only

The original cell-type classification for the chimera data is based on data integration between the entire chimera and the reference atlas. We now refine this by considering only the cells assigned to the limb mesoderm lineage. 

```{r}
Celltype <- "Limb_mesoderm"
atlas_meta <- readRDS("../fromIvan_new_atlas/integrated_meta_celltype_clus.rds")
atlas_sce <- readRDS("../data/big_atlas/big_atlas.rds")
cor_T <- readRDS(paste0("results_no_split/",Celltype,"/",Celltype,"_cor_cells_T.rds"))
cor_WT <- readRDS(paste0("results_no_split/",Celltype,"/",Celltype,"_cor_cells_WT.rds"))
correlation_pseudotime_celltypes_T <- atlas_meta$celltype.clustering[match(names(correlation_pseudotime_T),colnames(atlas_sce))]

chimeraT <- refine_celltype_stage(chimeraT,cor_T)
chimeraWT <- refine_celltype_stage(chimeraWT,cor_WT)
```


Plot pseudotime for limb mesoderm trajectory and plotting celltypes and T expression along it

```{r diffusion_limb_mesoderm}
diffusion_map_limb_mesoderm <- readRDS(paste0("results_no_split/",Celltype,"/Limb_mesoderm_diffusion_map_sublineage_Lateral_plate_mesoderm.rds"))
stages <- atlas_meta$stage[match(names(diffusion_map_limb_mesoderm$DC1),atlas_meta$cell)]
stages_double <- as.double(substring(stages,2,10))
cors <- cor(diffusion_map_limb_mesoderm@eigenvectors,stages_double)
diffusion_map_limb_mesoderm@eigenvectors[,cors <0 ] <- -diffusion_map_limb_mesoderm@eigenvectors[,cors <0 ]

celltype <- atlas_meta$celltype.clustering[match(names(diffusion_map_limb_mesoderm$DC1),atlas_meta$cell)]
T_expression <- logcounts(atlas_sce)["ENSMUSG00000062327",match(names(diffusion_map_limb_mesoderm$DC1),atlas_meta$cell)]
xx <- sample(1:length(diffusion_map_limb_mesoderm$DC1))
tmp <- data.frame(DC1 = eigenvectors(diffusion_map_limb_mesoderm)[xx, 1],
                    DC2 = eigenvectors(diffusion_map_limb_mesoderm)[xx, 2],
                    celltype = celltype[xx],
                    dpt = eigenvectors(diffusion_map_limb_mesoderm)[xx, 1],
                  stage=stages[xx],
                  T_expression=T_expression[xx])
  p1 <- ggplot(tmp, aes(x = DC1, y = DC2, colour = celltype)) +
    geom_point(size=1,alpha=1) + 
    xlab("Diffusion component 1") + 
    ylab("Diffusion component 2") +
    theme_classic(base_size=16) + theme(legend.position = "bottom",legend.box="vertical",legend.margin=margin())+
    labs(color="")+ guides(color = guide_legend(nrow=3,override.aes = list(size = 3,alpha=1))) + scale_color_manual(values=celltype_colours_final[unique(tmp$celltype)])
  print(p1)
  
 p2 <- ggplot(tmp[order(tmp$T_expression,decreasing=FALSE),], aes(x = DC1, y = DC2, colour = T_expression)) +
  geom_point(size=1,alpha=1) + 
  xlab("Diffusion component 1") + 
  ylab("Diffusion component 2") +
  theme_classic(base_size=16) + theme(legend.position = "bottom",legend.box="vertical",legend.margin=margin())+
  labs(color="") +scale_color_viridis_c(option="plasma") 
print(p2)
```
Alternative plot for T expression along pseudotime to make it in line with Mixl1 plot

```{r}
p4 <- ggplot(tmp[order(tmp$T_expression),], aes(x = DC1, y =  T_expression)) +
    geom_point(size=3,alpha=0.001)+
    xlab("Diffusion component 1") + 
    ylab("T expression (log2)") + geom_smooth(size=3)+ geom_point(data=tmp,mapping=aes(x=DC1,y=T_expression,color=stage),size=3,alpha=0.3)+
    theme_classic(base_size=20) +scale_color_viridis_d()+ theme(legend.position = "bottom",legend.box="vertical",legend.margin=margin())+
    labs(color="") + guides(color = guide_legend(override.aes = list(size = 6,alpha=1) ) )
  print(p4)
```

Plotting histogramme of celltypes along diffusion component 1

```{r histograms_diffusion_limb_mesoderm,fig.height=4,fig.width=7}
 p1 <- ggplot(tmp, aes(x = DC1,fill = celltype)) +
    geom_histogram(alpha=1) + 
    xlab("Diffusion component 1") +
    theme_classic(base_size=16) + theme(legend.position = "bottom",legend.box="vertical",legend.margin=margin())+
    labs(color="")+ guides(fill = guide_legend(nrow=3,override.aes = list(size = 3,alpha=1)))+ scale_fill_manual(values=celltype_colours_final[unique(tmp$celltype)])
  print(p1)
  
  
```
Differential expression 

```{r }
pseudotime_case <- correlation_pseudotime_T[["Limb_mesoderm"]]
pseudotime_control <-correlation_pseudotime_WT[["Limb_mesoderm"]]
gene_conv <-  read.table("../data/genes.tsv")
Celltype <- "Limb_mesoderm"
gene_file <- list.files(paste0("results_no_split/",Celltype),pattern="genes_sublineage")
temporal_genes <- read.table(paste0("results_no_split/",Celltype,"/",gene_file))$V1
temporal_genes <- gene_conv$V2[match(temporal_genes,gene_conv$V1)]

```

Differential expression per cell type along the lineage

```{r static_limb_mesoderm_per_celltype}
DE_genes_per_celltype <- list()
celltypes_along_trajectory <- unique(tmp$celltype)
for (j in 1:length(celltypes_along_trajectory)){
  sce_case <- chimeraT[gene_conv$V1[gene_conv$V2 %in% temporal_genes],!(is.na(chimeraT$celltype.refined))]
  sce_case <- sce_case[,sce_case$celltype.refined==celltypes_along_trajectory[j]]
  sce_case$perturbed <- sce_case$tomato
  sce_control <- chimeraWT[,!(is.na(chimeraWT$celltype.refined))]
  sce_control <- sce_control[,sce_control$celltype.refined==celltypes_along_trajectory[j]]
  sce_control$perturbed <- sce_control$tomato
  if (ncol(sce_case)>=20 & ncol(sce_control)>=20){
    pseudotime_case <- correlation_pseudotime_T[["limb_mesoderm"]]
    pseudotime_control <-correlation_pseudotime_WT[["limb_mesoderm"]]
    sce_case <- logNormCounts(sce_case,size_factors=sce_case$sizeFactor)
    sce_control <- logNormCounts(sce_control,size_factors=sce_control$sizeFactor)
    gene_conv <-  read.table("../data/genes.tsv")
    rownames(sce_case) <- gene_conv$V2[match(rownames(sce_case),gene_conv$V1)]
    rownames(sce_control) <- gene_conv$V2[match(rownames(sce_control),gene_conv$V1)]
    
    DE_expression_static_T_limb_mesoderm <- DE_expression_static(sce_case[temporal_genes,],sce_control[temporal_genes,])
    
    print(volcano_plot_static(DE_expression_static_T_limb_mesoderm,FDR_static=0.1,max_highlight=25) + ggtitle(paste0("T: ",celltypes_along_trajectory[j])))
    
  df_static <- DE_expression_static_T_limb_mesoderm[,c("static_effect.gene","static_effect_contrasted_with_control.estimate","static_effect_contrasted_with_control.FDR","static_effect.estimate", "static_effect.FDR")]
    colnames(df_static) <- c("gene","lfc_contrasted_with_control","FDR_contrast","lfc_case_only","FDR_case_only")
    write.table(df_static[df_static$FDR_contrast <0.1&df_static$FDR_case_only<0.1,], file=paste0("DE_expression_static_T_limb_mesoderm_",celltypes_along_trajectory[j],".csv"),sep=",",row.names=TRUE,col.names=TRUE)
  write.table(df_static, file=paste0("DE_expression_static_T_limb_mesoderm_all_",celltypes_along_trajectory[j],".csv"),sep=",",row.names=TRUE,col.names=TRUE)
  }
  DE_genes_per_celltype[[j]] <- df_static[df_static$FDR_contrast <0.1&df_static$FDR_case_only<0.1,]$gene
}
names(DE_genes_per_celltype) <- celltypes_along_trajectory
```
Illustration with boxplot

```{r}
chimeraWT <- logNormCounts(chimeraWT,size.factors=chimeraWT$sizeFactor)
chimeraT <- logNormCounts(chimeraT,size.factors=chimeraT$sizeFactor)

sce_case <- chimeraT[gene_conv$V1[gene_conv$V2 %in% DE_genes_per_celltype[["Lateral plate mesoderm"]]],!(is.na(chimeraT$celltype.refined))]
sce_case <- sce_case[,sce_case$celltype.refined=="Lateral plate mesoderm"]

sce_control <- chimeraWT[gene_conv$V1[gene_conv$V2 %in% DE_genes_per_celltype[["Lateral plate mesoderm"]]],!(is.na(chimeraWT$celltype.refined))]
sce_control <- sce_control[,sce_control$celltype.refined=="Lateral plate mesoderm"] 
  
logcounts_merged <- cbind(logcounts(sce_case),logcounts(sce_control))
rownames(logcounts_merged) <- gene_conv$V2[match(rownames(logcounts_merged),gene_conv$V1)]

tomato_T_WT <- c(ifelse(sce_case[,!(is.na(sce_case$celltype.refined))]$tomato,"T tomato pos","T tomato neg"),
                 ifelse(sce_control[,!(is.na(sce_control$celltype.refined))]$tomato,"WT tomato pos","WT tomato neg"))
df <- data.frame(gex=as.vector(logcounts_merged),tomato=as.factor(rep(tomato_T_WT,each=nrow(logcounts_merged))),
                 gene=rownames(logcounts_merged),ncol(logcounts_merged))
  
ggplot(df,aes(x=gene,color=tomato,y=gex)) +geom_boxplot()+ guides(colour = guide_legend(override.aes = list(alpha = 1,size=2))) +
  scale_color_manual(values=c("T tomato pos"="darkred","T tomato neg"="darkblue", "WT tomato pos"="orange", "WT tomato neg"="lightblue"))+theme_bw(base_size=8)+xlab("")+ylab("")+ theme(axis.text=element_text(size=10))+theme(legend.position="None")
```

# Pathway analysis

First we convert mouse genes to human orthologs.

```{r}
library(babelgene)
ortho <- orthologs(genes = rownames(chimeraT), species = "mouse", human = FALSE)
```



Pathway analysis with MAYA

```{r MAYA_limb_mesoderm_per_celltype}
rowData(chimeraT)$human_ortholog <- ortho$human_symbol[match(rownames(chimeraT),ortho$ensembl)]
rowData(chimeraWT)$human_ortholog <- ortho$human_symbol[match(rownames(chimeraWT),ortho$ensembl)]
counts_T <- counts(chimeraT[,!(is.na(chimeraT$celltype.refined))])
colnames(counts_T) <- paste0(colnames(counts_T),"_T")
rownames(counts_T) <- rowData(chimeraT)$human_ortholog 
perturbed_T <- chimeraT[,!(is.na(chimeraT$celltype.refined))]$tomato
celltype_T <- chimeraT[,!(is.na(chimeraT$celltype.refined))]$celltype.refined

counts_WT <- counts(chimeraWT[,!(is.na(chimeraWT$celltype.refined))])
rownames(counts_WT) <- rowData(chimeraWT)$human_ortholog 
perturbed_WT <- chimeraWT[,!(is.na(chimeraWT$celltype.refined))]$tomato
celltype_WT <- chimeraWT[,!(is.na(chimeraWT$celltype.refined))]$celltype.refined
colnames(counts_WT) <- paste0(colnames(counts_WT),"_WT")
MAYA_T_WT <- MAYA_pathway_analysis(expr_mat=cbind(counts_T,counts_WT),
                                        modules_list = "hallmark",
                                        is_logcpm=F)
MAYA_activity_matrix <- MAYA_T_WT$activity_matrix
activity_T <- MAYA_activity_matrix[,colnames(counts_T)]
activity_WT <- MAYA_activity_matrix[,colnames(counts_WT)] 

celltypes_along_trajectory <- c("Lateral plate mesoderm","Limb mesoderm")
for (j in 1:length(celltypes_along_trajectory)){
  sce_case <- SingleCellExperiment(assays=list(logcounts=activity_T[, celltype_T==celltypes_along_trajectory[j]]),
                                   colData=list(perturbed=perturbed_T[celltype_T==celltypes_along_trajectory[j]]))
  sce_control <- SingleCellExperiment(assays=list(logcounts=activity_WT[, celltype_WT==celltypes_along_trajectory[j]]),
                                   colData=list(perturbed=perturbed_WT[celltype_WT==celltypes_along_trajectory[j]]))
  if (ncol(sce_case)>=20 & ncol(sce_control)>=20){
  
    DE_expression_static_T_limb_mesoderm <- DE_expression_static(sce_case,sce_control)
    
    print(volcano_plot_static(DE_expression_static_T_limb_mesoderm,FDR_static=0.1,max_highlight=25) + ggtitle(paste0("T: ",celltypes_along_trajectory[j])))
    
 
  }
}
```


```{r hallmark_sets_geneDE}
hallmark_sets <- names(MAYA_T_WT$PCA_obj)
for (j in 1:length(hallmark_sets)){
  gene_contributions <- MAYA_T_WT$PCA_obj[[j]]$gene_contrib
  for (k in 1:nrow(gene_contributions)){
     gene_contributions_2 <- gene_contributions[k,]
     names(gene_contributions_2) <- ortho$symbol[match(names(gene_contributions_2),ortho$human_symbol)]
     gene_contributions_1 <- gene_contributions_2[order(gene_contributions_2,decreasing=TRUE)[1:min(50,length(gene_contributions_2))]]
    
  pp2 <- ggplot(mapping=aes(x=factor(names(gene_contributions_1),levels=names(gene_contributions_1)),y=gene_contributions_1,fill=gene_contributions_1>0)) + geom_bar(stat="identity")+theme_classic()+ggtitle(paste0(hallmark_sets[j],"-mode ",k))+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))+xlab("")+ylab("PC weight of gene")+
    scale_fill_manual(values=c("TRUE"="blue","FALSE"="red"))+theme(legend.position = "none")
  print(pp2)
  }
 
}
```

# Expression of key DE genes and pathways along atlas trajectory

```{r aldh2_expr}
Aldh2_expression <- logcounts(atlas_sce)["ENSMUSG00000029455",match(names(diffusion_map_limb_mesoderm$DC1),atlas_meta$cell)]
xx <- sample(1:length(diffusion_map_limb_mesoderm$DC1))
tmp <- data.frame(DC1 = eigenvectors(diffusion_map_limb_mesoderm)[xx, 1],
                    DC2 = eigenvectors(diffusion_map_limb_mesoderm)[xx, 2],
                    celltype = celltype[xx],
                    dpt = eigenvectors(diffusion_map_limb_mesoderm)[xx, 1],
                  stage=stages[xx],
                  Aldh2_expression=Aldh2_expression[xx])
 
 p2 <- ggplot(tmp, aes(x = DC1, y = Aldh2_expression)) +
  geom_point(size=2,alpha=0.2) + 
  xlab("pseudotime") + 
  ylab("Aldh2 expression") +
  theme_classic(base_size=16) + theme(legend.position = "bottom",legend.box="vertical",legend.margin=margin())+
  labs(color="") +scale_color_viridis_c(option="plasma") + geom_smooth()
print(p2)
```

```{r vim_expr}
Vim_expression <- logcounts(atlas_sce)["ENSMUSG00000026728",match(names(diffusion_map_limb_mesoderm$DC1),atlas_meta$cell)]
xx <- sample(1:length(diffusion_map_limb_mesoderm$DC1))
tmp <- data.frame(DC1 = eigenvectors(diffusion_map_limb_mesoderm)[xx, 1],
                    DC2 = eigenvectors(diffusion_map_limb_mesoderm)[xx, 2],
                    celltype = celltype[xx],
                    dpt = eigenvectors(diffusion_map_limb_mesoderm)[xx, 1],
                  stage=stages[xx],
                  Vim_expression=Vim_expression[xx])
 
 p2 <- ggplot(tmp, aes(x = DC1, y = Vim_expression)) +
  geom_point(size=2,alpha=0.2) + 
  xlab("pseudotime") + 
  ylab("Vim expression") +
  theme_classic(base_size=16) + theme(legend.position = "bottom",legend.box="vertical",legend.margin=margin())+
  labs(color="") +scale_color_viridis_c(option="plasma") + geom_smooth()
print(p2)
```

```{r Apela_expr}
Apela_expression <- logcounts(atlas_sce)["ENSMUSG00000079042",match(names(diffusion_map_limb_mesoderm$DC1),atlas_meta$cell)]
xx <- sample(1:length(diffusion_map_limb_mesoderm$DC1))
tmp <- data.frame(DC1 = eigenvectors(diffusion_map_limb_mesoderm)[xx, 1],
                    DC2 = eigenvectors(diffusion_map_limb_mesoderm)[xx, 2],
                    celltype = celltype[xx],
                    dpt = eigenvectors(diffusion_map_limb_mesoderm)[xx, 1],
                  stage=stages[xx],
                  Apela_expression=Apela_expression[xx])
 
 p2 <- ggplot(tmp, aes(x = DC1, y = Apela_expression)) +
  geom_point(size=2,alpha=0.2) + 
  xlab("pseudotime") + 
  ylab("Apela expression") +
  theme_classic(base_size=16) + theme(legend.position = "bottom",legend.box="vertical",legend.margin=margin())+
  labs(color="") +scale_color_viridis_c(option="plasma") + geom_smooth()
print(p2)
```