---
title: "AML data analysis"
output:
   BiocStyle::html_document:
    code_folding: hide
    number_sections: yes
    toc: yes  
---

This script applies COSICC to published scRNA-seq data from AML patients and healthy donors (van Galen et al. (2019). Single-Cell RNA-Seq Reveals AML Hierarchies Relevant to Disease Progression and Immunity. Cell.)                         

# Data loading and computation of percentage of malignant cells 

First, we read in the data sets and save them as a list of SingleCellExperiments, named by donor and time point. 

```{r}
suppressPackageStartupMessages(source("../core_functions_dynamic.R"))
suppressPackageStartupMessages(source("../core_functions_DA_testing_general.R"))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(Polychrome))
suppressPackageStartupMessages(library(ggrepel))
suppressPackageStartupMessages(library(COSICC))
```

```{r}
set.seed(44)
knitr::opts_chunk$set(
    echo = FALSE,
    message = FALSE,
    warning = FALSE,
    message = FALSE
)
options(datatable.fread.datatable=FALSE)
suppressPackageStartupMessages(library(SingleCellExperiment))
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(destiny))
suppressPackageStartupMessages(library(scran))
suppressPackageStartupMessages(library(scater))
suppressPackageStartupMessages(library(batchelor))
```

```{r, eval=TRUE}
file_names_anno <- list.files("vanGalen_data",pattern="anno")
file_names_dem <- list.files("vanGalen_data",pattern="dem")
sample_names <- sapply(file_names_anno,function(x) strsplit(x,"_")[[1]][2])
sample_names <- sapply(sample_names,function(x) strsplit(x,"[.]")[[1]][1])

sce_list <- list()
for (j in 1:length(sample_names)){

  data_dem <- fread(paste0("vanGalen_data/",file_names_dem[grepl(paste0(sample_names[j],".dem"),file_names_dem)]),header=TRUE)
  data_anno <- fread(paste0("vanGalen_data/",file_names_anno[grepl(paste0(sample_names[j],".anno"),file_names_anno)]),fill=TRUE)
  rownames(data_anno) <- data_anno$Cell
  rownames(data_dem) <- data_dem$Gene
  data_dem <- data_dem[,colnames(data_dem)!="Gene"]
  xx <- intersect(rownames(data_anno),colnames(data_dem))
  sce_list[[j]] <- SingleCellExperiment(assay=list(counts=data_dem[,xx]),colData=as.matrix(data_anno)[xx,])
}
sample_ind_patient <- grepl("D",sample_names)
days <- sapply(sample_names[sample_ind_patient],function(x) strsplit(x,"D")[[1]][2])
add_zeros <- function(x){
  if(nchar(x) ==2){
    x <- paste0("0",x)
  }
  if (nchar(x) == 1){
    x <- paste0("00",x)
  }
  return(x)
}
days <- sapply(days,add_zeros)
names(sce_list) <- sample_names
sample_names[sample_ind_patient] <- paste0(sapply(sample_names[sample_ind_patient],function(x) strsplit(x,"D")[[1]][1]),
                                              "D",days)
names(sce_list) <- sample_names
saveRDS(sce_list,"sce_list_AML.rds")
```

Now we compute the proportion of malignant cells for each of the patient data sets, for the myeloid trajectory. 

```{r proportion_perturbed}
sce_list <- readRDS("sce_list_AML.rds")
sce_list <- sapply(sce_list,function(x) return(x[,x$PredictionRefined%in%c("malignant","normal")]))
sce_list <- sapply(sce_list,function(x) {x$perturbed = x$PredictionRefined == "malignant";return(x)})
sce_list <- sapply(sce_list,function(x) {names(colData(x))[names(colData(x))=="CellType"] = "celltype";return(x)})
sce_list <- sapply(sce_list,function(x) {x$celltype = sapply(x$celltype,function(y) strsplit(y,"-")[[1]][1]);return(x)})

saveRDS(sce_list,"sce_list_AML.rds")
```


```{r proportion_perturbed_myeloid, fig.height=2.5,fig.width=7}
sce_list_my <- sapply(sce_list,function(x) x[,x$celltype %in% c("HSC","Prog","GMP","ProMono","Mono")])
proportion_perturbed <- sapply(sce_list_my,function(x) sum(x$perturbed/ncol(x)))
names(proportion_perturbed) <- names(sce_list)
proportion_perturbed <- proportion_perturbed[grepl("-D",names(proportion_perturbed))]
proportion_perturbed <- proportion_perturbed[proportion_perturbed>0]
patient=sapply(names(proportion_perturbed),function(x) strsplit(x,"-")[[1]][1])
patients_keep <- names(table(patient)[table(patient) >1])
proportion_perturbed <- proportion_perturbed[patient%in%patients_keep]
patient <- patient[patient%in%patients_keep]

patients_keep <- patient[proportion_perturbed[paste0(patient,"-D000")] < 0.95]
proportion_perturbed <- proportion_perturbed[patient%in%patients_keep]
patient <- patient[patient%in%patients_keep]

glasbey_cols=glasbey.colors(length(unique(patient))+2)
glasbey_cols <- glasbey_cols[2:length(glasbey_cols)]
glasbey_cols <- glasbey_cols[-3]
names(glasbey_cols) <- unique(patient)
x_lab_colours <- glasbey_cols[patient]
ggplot(mapping=aes(x=names(proportion_perturbed),y=proportion_perturbed,fill=patient)) + geom_bar(stat="identity")+theme_classic()+ theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))+ xlab("")+scale_fill_manual(values= glasbey_cols)+guides(fill = guide_legend(ncol=2,override.aes = list(size = 3)))+theme(axis.text.x = element_text(color=x_lab_colours))
saveRDS(proportion_perturbed,file="proportion_perturbed_cells_AML.rds")
```

# Pseudotime for controls datasets

First, we merge the control data sets to one obtain one SingleCellExperiment data set. 

```{r}
sce_list <- readRDS("sce_list_AML.rds")
sce_list_control <- sce_list[names(sce_list)[grepl("BM",names(sce_list))]]
for (j in 1:length(sce_list_control)){
  sce_list_control[[j]]$batch <- names(sce_list_control)[j]
}
sce_list_control <- multiBatchNorm(sce_list_control)
sce_control <- do.call(cbind,sce_list_control)
```

We subset this SingleCellExperiment to cell types of the myeloid trajectory.

```{r}
sce_control_mt <- sce_control[,sce_control$celltype %in% c("HSC","Prog","GMP","ProMono","Mono")]

```

## Myeloid-trajectory genes

We identify genes that are DE across cell types but not across batches. As there are no time points, we use the cell types instead of stage and identify cells that are DE across cell types to use for pseudotime. 

```{r}
sce_control_mt$stage <- NA
sce_control_mt$stage[sce_control_mt$celltype == "HSC"]  <- 1
sce_control_mt$stage[sce_control_mt$celltype == "Prog"]  <- 2
sce_control_mt$stage[sce_control_mt$celltype == "GMP"]  <- 3
sce_control_mt$stage[sce_control_mt$celltype == "ProMono"]  <- 4
sce_control_mt$stage[sce_control_mt$celltype == "Mono"]  <- 5
sce_control_mt_dynamic_genes <- find_dynamic_genes(sce_control_mt)
```

We plot diffusion trajectories coloured by pseudotime and by cell types. 

```{r}
sce_control_mt_dynamic_genes <- compute_and_plot_pseudotime(sce_control_mt_dynamic_genes)
```

Now we map the disease samples to the reference pseudotime using correlation-based mapping on the genes identified above as varying across stage and not batch. 

```{r}
sce_list_AML <- sce_list[names(proportion_perturbed)[proportion_perturbed < 0.95 & proportion_perturbed > 0.05]]
for (j in 1:length(sce_list_AML)){
  sce_list_AML[[j]]$batch <- names(sce_list_AML)[j]
  colData(sce_list_AML[[j]]) <- colData(sce_list_AML[[j]])[setdiff(names(colData(sce_list_AML[[j]])),"NanoporeTranscripts")]
}
sce_list_AML <- multiBatchNorm(sce_list_AML)
sce_AML <- do.call(cbind,sce_list_AML)
celltypes_keep <- c(unique(sce_control_mt_dynamic_genes$CellType),paste0(unique(sce_control_mt_dynamic_genes$CellType),"-like"))
sce_AML <- sce_AML[,sce_AML$CellType%in%celltypes_keep]
sce_AML_dynamic_genes <- sce_AML[rownames(sce_control_mt_dynamic_genes),]
sce_list <- lapply(sce_list,logNormCounts)
sce_list_patients <-  sce_list[grepl("AML",names(sce_list))]
sce_list_patients <- sce_list_patients[intersect(names(sce_list_patients),
                              names(proportion_perturbed)[proportion_perturbed > 0.1])]
sce_list_dynamic_genes <- sapply(sce_list,function(x) return(x[rownames(sce_control_mt_dynamic_genes),]))
sce_list_dynamic_genes_AML <- sce_list_dynamic_genes[grepl("AML",names(sce_list_dynamic_genes))]
sce_list_dynamic_genes_AML <- sce_list_dynamic_genes_AML[intersect(names(sce_list_dynamic_genes_AML),
                              names(proportion_perturbed)[proportion_perturbed > 0.1])]
```

# COSICC_kinetics

Now we perform COSICC_kinetics to test whether there is delay or faster progression along the myeoloid trajectory for the malignant compared to the normal cells. 

```{r, eval=TRUE}
wilcox_AML <- list()
correlation_matrices <- list()
for (j in 1:length(sce_list_dynamic_genes_AML)){
  temp <- correlation_pseudotime(reference_sce=sce_control_mt_dynamic_genes,perturbed_sce=sce_list_dynamic_genes_AML[[j]])
  sce_AML1 <- sce_list_dynamic_genes_AML[[j]]
  colData(sce_AML1) <- colData(temp$perturbed_sce)
  correlation_matrices[[j]] <- temp$correlation_matrix
  sce_AML1$perturbation_status <- sce_AML1$PredictionRefined
  sce_AML1 <- sce_AML1[,sce_AML1$perturbation_status%in%c("normal","malignant")]
  sce_AML1$perturbation_status[sce_AML1$PredictionRefined=="malignant"] <- "perturbed"
  if (length(unique(sce_AML1$perturbation_status)) > 1){
    wilcox_AML[[j]] <- Wilcoxon_test_perturbed_vs_normal(sce_AML1)
    sce_list_dynamic_genes_AML[[j]] <- sce_AML1
  }
}
names(wilcox_AML) <- names(sce_list_dynamic_genes_AML)[1:length(wilcox_AML)]
wilcox_AML <- wilcox_AML[unlist(lapply(wilcox_AML,function(x) length(x) > 0))]
for (j in 1:length(wilcox_AML)){
  wilcox_AML[[j]]$sample <- names(wilcox_AML)[j]
}
wilcox_AML <- do.call(rbind,wilcox_AML)
wilcox_AML <- wilcox_AML[order(wilcox_AML$sample),]
wilcox_AML$FDR <- p.adjust(wilcox_AML$p_value)
wilcox_AML$sig <- wilcox_AML$FDR < 0.1
saveRDS(wilcox_AML,file="wilcox_AML.rds")
saveRDS(sce_list_dynamic_genes_AML,file="sce_list_dynamic_genes_AML.rds")
```

We plot the Wilcoxon estimate and confidence interval for D0 samples. 

```{r wilcox_D0,fig.height=2,fig.width=5}
wilcox_AML <- readRDS("wilcox_AML.rds")
wilcox_AML_D0 <- wilcox_AML[grepl("D0",wilcox_AML$sample),]
patient_wilcox_D0 <- sapply(wilcox_AML_D0$sample,function(x) return(strsplit(x,"-")[[1]][1]))
colours_AML_D0 <- glasbey_cols[patient_wilcox_D0]
colours_AML_D0[!(wilcox_AML_D0$sig)] <- "grey"
pp <- ggplot(wilcox_AML_D0, aes(x = lower_bound, y = sample)) +
  geom_segment(aes(xend = upper_bound, yend = sample),size=1,color=colours_AML_D0) + 
geom_point(size=2,color=colours_AML_D0) + 
geom_point(aes(x = upper_bound),size=2,color=colours_AML_D0)  +
  xlab("conf. intervall\nWilcoxon statistic") + theme_bw() + geom_vline(xintercept = 0)+
theme(axis.text.y = element_text(color=glasbey_cols[patient_wilcox_D0]))
print(pp)
```

Plotting Wilcoxon estimate and confidence interval for patients 328, 329, and 707B. 

```{r wilcox_per_patient,fig.height=2,fig.width=5}
wilcox_AML_sub <- wilcox_AML[wilcox_AML$sample%in%c("AML328-D000","AML328-D029","AML329-D000","AML329-D200","AML707B-D000","AML707B-D108"),]
patients <- c("AML328","AML329","AML707B")
for (j in 1:length(patients)){
    wilcox_AML_patient <- wilcox_AML_sub[grepl(patients[j],rownames(wilcox_AML_sub)),]
    colours_AML <- rep(glasbey_cols[patients[j]],nrow(wilcox_AML_patient))
    colours_AML[!(wilcox_AML_patient$sig)] <- "grey"
    pp <- ggplot(wilcox_AML_patient, aes(x = lower_bound, y = sample)) +
    geom_segment(aes(xend = upper_bound, yend = sample),size=1,color=colours_AML) + 
  geom_point(size=2,color=colours_AML)+
  geom_point(aes(x = upper_bound),size=2,color=colours_AML)  +
    xlab("conf. intervall\nWilcoxon statistic") + theme_bw() + geom_vline(xintercept = 0)
  print(pp)
}
```


```{r}
sessionInfo()
```